File: src/discord/renderers/diffView.ts
import type { AgentEvent } from '../../core/contracts.js';
import { chunkMessage } from '../format.js';
import type { TurnThreadRegistry } from './turnThread.js';

// File-change diff view (§6, §5a — Claude, cap fileDiff): when a file-editing tool
// (Edit/Write/NotebookEdit) completes successfully, post a diff of the change into the
// thread resolved by TurnThreadRegistry (main or the subagent thread that owns the edit).
// The diff must be built from the tool INPUT (old/new strings), which lives on the
// `tool_use` event, not the `tool_result` — so this handler tracks edit inputs by id and
// renders on the matching successful result (mirrors A4D's changeTracker → diffViewer
// flow). No discord.js: the sink is the MessageThread port (via TurnThreadRegistry).

// The tools whose input this handler renders as a pretty diff. Exported so
// ToolThreadHandler skips their raw input (rendering it too would duplicate the change).
// `apply_patch` is Codex app-server fileChange (changes[] with unified diffs).
export const FILE_EDIT_TOOLS = new Set(['Edit', 'Write', 'NotebookEdit', 'apply_patch']);

interface EditInput {
  filePath: string;
  oldString?: string;
  newString?: string;
  content?: string;
  // Pre-formatted unified diff body (Codex apply_patch). When set, renderDiff uses it as-is.
  unifiedDiff?: string;
}

export interface DiffViewDeps {
  registry: TurnThreadRegistry;
}

export class DiffViewHandler {
  private readonly registry: TurnThreadRegistry;
  // toolUseId → the edit input, kept until the matching result arrives.
  private readonly edits = new Map<string, EditInput>();

  constructor(deps: DiffViewDeps) {
    this.registry = deps.registry;
  }

  // Record a file-edit tool_use so its result can be diffed. Non-edit tools are
  // ignored (nothing tracked → nothing rendered). Also binds the tool to its thread
  // key so handleResult can open the right thread even if toolThreads is off.
  noteToolUse(ev: Extract<AgentEvent, { kind: 'tool_use' }>): void {
    if (!FILE_EDIT_TOOLS.has(ev.name)) return;
    const parsed = parseEditInput(ev.input);
    if (parsed) this.edits.set(ev.id, parsed);
    this.registry.bindToolUse(ev);
  }

  // On a successful result for a tracked edit, post its diff. A failed result or an
  // untracked id renders nothing.
  async handleResult(ev: Extract<AgentEvent, { kind: 'tool_result' }>): Promise<void> {
    const edit = this.edits.get(ev.id);
    if (!edit) return;
    this.edits.delete(ev.id);
    if (!ev.ok) return;
    const diff = renderDiff(edit);
    if (!diff) return;
    // Always open the resolved thread (diff posts even when toolThread would buffer).
    const key = this.registry.resolveResultKey(ev);
    const thread = await this.registry.get(key);
    for (const chunk of chunkMessage('```diff\n' + diff + '\n```')) {
      await thread.send({ content: chunk });
    }
  }
}

// Build a unified-ish diff body (no headers) from an edit input. Write/create with
// only `content` renders all lines as additions; an Edit renders old→new.
// Codex apply_patch supplies a pre-formatted unifiedDiff string.
function renderDiff(edit: EditInput): string | null {
  if (edit.unifiedDiff !== undefined && edit.unifiedDiff.length > 0) {
    return edit.unifiedDiff;
  }
  const header = `--- ${edit.filePath}`;
  if (edit.oldString !== undefined || edit.newString !== undefined) {
    const removed = (edit.oldString ?? '').split('\n').map((l) => `- ${l}`);
    const added = (edit.newString ?? '').split('\n').map((l) => `+ ${l}`);
    return [header, ...removed, ...added].join('\n');
  }
  if (edit.content !== undefined) {
    const added = edit.content.split('\n').map((l) => `+ ${l}`);
    return [header, ...added].join('\n');
  }
  return null;
}

function parseEditInput(input: unknown): EditInput | null {
  if (typeof input !== 'object' || input === null || Array.isArray(input)) return null;
  const obj = input as Record<string, unknown>;

  // Codex apply_patch: { changes: [{ path, kind, diff }, ...] }
  if (Array.isArray(obj.changes)) {
    const parts: string[] = [];
    for (const c of obj.changes) {
      if (typeof c !== 'object' || c === null || Array.isArray(c)) continue;
      const rec = c as Record<string, unknown>;
      const pathStr = typeof rec.path === 'string' ? rec.path : '';
      const diff = typeof rec.diff === 'string' ? rec.diff : '';
      if (diff.length > 0) {
        parts.push(pathStr.length > 0 ? `--- ${pathStr}\n${diff}` : diff);
      } else if (pathStr.length > 0) {
        parts.push(pathStr);
      }
    }
    if (parts.length === 0) return null;
    const first = obj.changes[0];
    const firstPath =
      typeof first === 'object' &&
      first !== null &&
      typeof (first as { path?: unknown }).path === 'string'
        ? (first as { path: string }).path
        : 'patch';
    return { filePath: firstPath, unifiedDiff: parts.join('\n\n') };
  }

  const filePath =
    typeof obj.file_path === 'string'
      ? obj.file_path
      : typeof obj.path === 'string'
        ? obj.path
        : undefined;
  if (!filePath) return null;
  const out: EditInput = { filePath };
  if (typeof obj.old_string === 'string') out.oldString = obj.old_string;
  else if (typeof obj.oldText === 'string') out.oldString = obj.oldText;
  if (typeof obj.new_string === 'string') out.newString = obj.new_string;
  else if (typeof obj.newText === 'string') out.newString = obj.newText;
  if (typeof obj.content === 'string') out.content = obj.content;
  return out;
}

File: swift/Sources/DiscordAgentBridge/Render/DiffView.swift
import Foundation

// File-change diff view (TS `src/discord/renderers/diffView.ts`).
// Pure parse/render + handler that posts into TurnThreadRegistry.

/// Tools whose input this handler renders as a pretty diff. ToolThreadHandler skips their raw input.
public let FILE_EDIT_TOOLS: Set<String> = ["Edit", "Write", "NotebookEdit", "apply_patch"]

public struct EditInput: Sendable, Equatable {
    public var filePath: String
    public var oldString: String?
    public var newString: String?
    public var content: String?
    /// Pre-formatted unified diff body (Codex apply_patch).
    public var unifiedDiff: String?

    public init(
        filePath: String,
        oldString: String? = nil,
        newString: String? = nil,
        content: String? = nil,
        unifiedDiff: String? = nil
    ) {
        self.filePath = filePath
        self.oldString = oldString
        self.newString = newString
        self.content = content
        self.unifiedDiff = unifiedDiff
    }
}

/// Build a unified-ish diff body (no headers) from an edit input.
public func renderDiff(_ edit: EditInput) -> String? {
    if let u = edit.unifiedDiff, !u.isEmpty { return u }
    let header = "--- \(edit.filePath)"
    if edit.oldString != nil || edit.newString != nil {
        let removed = (edit.oldString ?? "").split(separator: "\n", omittingEmptySubsequences: false).map { "- \($0)" }
        let added = (edit.newString ?? "").split(separator: "\n", omittingEmptySubsequences: false).map { "+ \($0)" }
        return ([header] + removed + added).joined(separator: "\n")
    }
    if let content = edit.content {
        let added = content.split(separator: "\n", omittingEmptySubsequences: false).map { "+ \($0)" }
        return ([header] + added).joined(separator: "\n")
    }
    return nil
}

/// Parse tool_use input into EditInput. Returns nil when no usable path/diff.
public func parseEditInput(_ input: JSONValue) -> EditInput? {
    guard let obj = input.objectValue else { return nil }

    // Codex apply_patch: { changes: [{ path, kind, diff }, ...] }
    if let changes = obj["changes"]?.arrayValue {
        var parts: [String] = []
        for c in changes {
            guard let rec = c.objectValue else { continue }
            let pathStr = rec["path"]?.stringValue ?? ""
            let diff = rec["diff"]?.stringValue ?? ""
            if !diff.isEmpty {
                parts.append(pathStr.isEmpty ? diff : "--- \(pathStr)\n\(diff)")
            } else if !pathStr.isEmpty {
                parts.append(pathStr)
            }
        }
        if parts.isEmpty { return nil }
        let firstPath: String = {
            if let first = changes.first?.objectValue, let p = first["path"]?.stringValue, !p.isEmpty {
                return p
            }
            return "patch"
        }()
        return EditInput(filePath: firstPath, unifiedDiff: parts.joined(separator: "\n\n"))
    }

    let filePath = obj["file_path"]?.stringValue ?? obj["path"]?.stringValue
    guard let filePath, !filePath.isEmpty else { return nil }
    var out = EditInput(filePath: filePath)
    if let s = obj["old_string"]?.stringValue { out.oldString = s }
    else if let s = obj["oldText"]?.stringValue { out.oldString = s }
    if let s = obj["new_string"]?.stringValue { out.newString = s }
    else if let s = obj["newText"]?.stringValue { out.newString = s }
    if let s = obj["content"]?.stringValue { out.content = s }
    return out
}

// MARK: - Handler

/// Records edit tool_use inputs and posts a ```diff``` on successful tool_result.
/// Mutable; intended for single-owner use (ToolActivityHost / tests).
public final class DiffViewHandler: @unchecked Sendable {
    private let registry: TurnThreadRegistry
    private var edits: [String: EditInput] = [:]

    public init(registry: TurnThreadRegistry) {
        self.registry = registry
    }

    public func noteToolUse(id: String, name: String, input: JSONValue, parentToolUseId: String?) {
        guard FILE_EDIT_TOOLS.contains(name) else { return }
        if let parsed = parseEditInput(input) {
            edits[id] = parsed
        }
        registry.bindToolUse(id: id, name: name, input: input, parentToolUseId: parentToolUseId)
    }

    public func handleResult(id: String, ok: Bool, content: String, parentToolUseId: String?) async {
        guard let edit = edits.removeValue(forKey: id) else { return }
        guard ok else { return }
        guard let diff = renderDiff(edit) else { return }
        let key = registry.resolveResultKey(id: id, parentToolUseId: parentToolUseId)
        do {
            let thread = try await registry.get(key)
            for chunk in DiscordText.chunkMessage("```diff\n\(diff)\n```") {
                try await thread.send(chunk)
            }
        } catch {
            // swallow — fire-and-forget like TS
        }
    }

    public func resetTurn() {
        edits.removeAll()
    }
}
